File: Java/src/trees/BinarySearchTree.java
package trees;

/**
 * Created by Chris Keyser on 3/11/2016.
 * Binary search tree implementation
 * Includes support for insertion, deletion, contains, height, and pre-order/in-order/post-order traversals
 */
public class BinarySearchTree {
    private BinarySearchTree left;
    private BinarySearchTree right;
    private BinarySearchTree parent;
    private int val;

    public BinarySearchTree(BinarySearchTree parent, int val) {
        this.val = val;
        this.parent = parent;
        left = null;
        right = null;
    }

    public int getVal() {
        return val;
    }

    /**
     * Inserts a value into the BST
     * @param i value to insert into the BST
     */
    public void insert(int i) {
        // if newVal is less than val, it should be added to the left side of the tree, otherwise add it to the right side
        if(i < val) {
            if(left == null) {
                //if there is no left child, create a new left child with newVal
                left = new BinarySearchTree(this, i);
            } else {
                // if there is a left child, add newVal to the tree on the left
                left.insert(i);
            }
        } else {
            if(right == null) {
                //if there is no right child, create a new right child with newVal
                right = new BinarySearchTree(this, i);
            } else {
                //if there is a right child, add newVal to the right side of the tree
                right.insert(i);
            }
        }
    }

    /**
     * determines whether the BST contains toFind
     * has efficiency of O(log(n)) because it at most needs to search up to the depth of the tree
     * @param i int to find in the BST
     * @return boolean
     */
    public boolean contains(int i) {
        if(i == val) {
            return true;
        } else if(i < val) {
            return left != null && left.contains(i);
        } else {
            return right != null && right.contains(i);
        }
    }

    /**
     * returns the height of the tree from the current node
     * if called on root will return the right of the whole tree
     * the tree can hold log(h) - 1 nodes
     * @return int
     */
    public int getHeight() {
        int childHeight = 0;

        //find the greater of this tree's children's lengths
        if(!this.isLeaf()) {
            int leftHeight = left != null ? left.getHeight() : 0;
            int rightHeight = right != null ? right.getHeight() : 0;

            childHeight = leftHeight > rightHeight ? leftHeight : rightHeight;
        }

        //return larger child tree length + 1 for current node
        return childHeight + 1;
    }

    /**
     * deletes the node containing i from the BST
     * this will break if you try to call it to delete the root of the tree
     * if the root of the tree has no children
     * @param i int value of node to delete
     */
    public void delete(int i) {
        //if the value we're looking to delete is less than the current node's value, look in the left subtree
        if(i < val) {
            if(left != null) {
                left.delete(i);
            }
        } else if(i > val) { // else look in the right subtree
            if(right != null) {
                right.delete(i);
            }
        } else { //else if not <> it must be this node so delete it
            if(this.isLeaf()) {
                //if there are no children just null the reference to this object in its parent
                parent.replaceChild(i, null);
            } else if(this.hasSingleChild()) {
                /**
                 * if there's a single child replace it with the child's descendant with the value closest to i
                 * meaning the smallest of right or the largest of left
                 */
                int descVal = left != null ? left.getLargestDescendant().getVal() : right.getSmallestDescendant().getVal();
                delete(descVal);
                val = descVal;
            } else {
                /**
                 * if there's two children replace this node with the largest descendant of its left child
                 * the leaf's value will be larger than all left subtree values and less than all right subtree values
                 * making its placement function equivalently to the current node
                 * deleting the largest descendant node will cascade until the tree resettles by replacing a value with
                 * a leaf
                 */
                int largestLeftDescVal = left.getLargestDescendant().getVal();
                delete(largestLeftDescVal);
                val = largestLeftDescVal;
            }
        }
    }

    /**
     * gets the largest of this tree's descendants
     * helper for delete()
     * @return BinarySearchTree
     */
    private BinarySearchTree getLargestDescendant() {
        if(this.isLeaf()) {
            return this;
        }

        //get the right child if one exists, otherwise look in left subtree
        BinarySearchTree child = this.right != null ? this.right : this.left;

        //find the rightmost descendant
        while(child.right != null) {
            child = child.right;
        }

        return child;
    }

    /**
     * gets the smallest of this tree's descendants
     * helper for delete()
     * @return BinarySearchTree
     */
    private BinarySearchTree getSmallestDescendant() {
        if(this.isLeaf()) {
            return this;
        }

        //get the right child if one exists, otherwise look in left subtree
        BinarySearchTree child = this.left != null ? this.left : this.right;

        //find the rightmost descendant
        while(child.left != null) {
            child = child.left;
        }

        return child;
    }

    /**
     * replaces the child with value childVal with BST newChild
     * @param childVal int
     * @param newChild BinarySearchTree
     */
    public void replaceChild(int childVal, BinarySearchTree newChild) {
        if(left != null && left.getVal() == childVal) {
            left = newChild;
        } else if(right != null && right.getVal() == childVal) {
            right = newChild;
        }
    }

    /**
     * determines if the current node has no children and is as such a leaf
     * @return bool
     */
    public boolean isLeaf() {
        return left == null && right == null;
    }

    /**
     * determines if this node has a single child
     * @return bool
     */
    public boolean hasSingleChild() {
        // ^ is XOR (exclusive or) operator, which is equivalent to (left != null || right != null) && !(left != null & right != null)
        return left != null ^ right != null;
    }

    /**
     * wrapper for pre-order traversal because I can't seem to get rid of leading whitespace otherwise
     */
    public String preOrder() {
        return preOrderHelper().trim();
    }

    /**
     * performs a pre-order traversal of the tree, returning a string representing the values in the order they were reached
     */
    public String preOrderHelper() {
        String leftTraversal = left != null ? left.preOrderHelper() : "";
        String rightTraversal = right != null ? right.preOrderHelper() : "";
        return val + " " + leftTraversal + rightTraversal;
    }

    /**
     * wrapper for in-order traversal because I can't seem to get rid of leading whitespace otherwise
     */
    public String inOrder() {
        String s = inOrderHelper();

        return s.trim();
    }

    /**
     * performs an in-order traversal of the tree, returning a string representing the values in the order they were reached
     */
    public String inOrderHelper() {
        String leftTraversal = left != null ? left.inOrderHelper() : "";
        String rightTraversal = right != null ? right.inOrderHelper() : "";

        return leftTraversal + " " + val + rightTraversal;
    }

    /**
     * wrapper for post-order traversal because I can't seem to get rid of leading whitespace otherwise
     */
    public String postOrder() {
        return postOrderHelper().trim();
    }

    /**
     * performs a post-order traversal of the tree, returning a string representing the values in the order they were reached
     */
    public String postOrderHelper() {
        String leftTraversal = left != null ? left.postOrderHelper() : "";
        String rightTraversal = right != null ? right.postOrderHelper() : "";
        return leftTraversal + " " + rightTraversal + val;
    }
}

File: js/src/trees/BinarySearchTree.js
function BinarySearchTreeNode(parent, val) {
  this.parent = parent;
  this.value = val;
  this.count = 1; // # of times this value has been inserted into the tree
  this.left = null;
  this.right = null;
}

// BST implementation
// I attempt to prevent invalid values from entering the tree
// by type-checking during insertion, but since I can't hide
// the instance variables it's possible to corrupt this if
// you access them directly instead of using the given functions
// to handle the tree
BinarySearchTreeNode.prototype = (function() {
  // insert given value into the tree, maintaining BST properties
  function insert(n) {
    if(typeof(n) !== 'number') {
      console.log('non-number value passed into BinarySearchTreeNode.insert()');
      return null;
    };

    if(n !== 0 && !n) {
      console.log('invalid value passed into BinarySearchTreeNode.insert(): ' + n);
      return null;
    }

    // if we already have this value in the tree, up the count on its node
    if(n === this.value) {
      this.count++;
      return this;
    }

    if(n < this.value) {
      if(this.left) {
        // if we have a left subtree, insert it into that subtree
        this.left.insert(n);
      } else {
        // if we don't, this is our new left child
        this.left = new BinarySearchTreeNode(this, n);
      }
    } else {
      if(this.right) {
        // if we have a right subtree, insert it into that subtree
        this.right.insert(n);
      } else {
        // if we don't, this is our new right child
        this.right = new BinarySearchTreeNode(this, n);
      }
    }
  }

  // remove node with value n from the tree, maintaining BST properties
  // if there are multiple, will remove the first found in an in-order traversal
  // returns a boolean representing whether the removal was successful
  function remove(n) {
    if(typeof(n) !== 'number') {
      console.log('non-number value passed into BinarySearchTreeNode.remove()');
      return false;
    };

    if(n != 0 && !n) {
      console.log('invalid value passed into BinarySearchTreeNode.remove(): ' + n);
      return false;
    }

    // if we've found the value
    if(n === this.value) {
      if(this.count > 1) {
        // if there's multiple of the value, remove one
        this.count--;
      } else {
        // otherwise get rid of the node entirely
        deleteNode(this);
      }
    } else if (n < this.value) {
      if(n.left) {
        // if it's in the left subtree, remove it from that
        n.left.remove(n);
      } else {
        // if it's less than this node's value but there's no left subtree, we don't have it
        console.log('value ' + n + ' was attempted to be removed from a tree that does not contain it');
        return false;
      }
    } else {
      if(n.right) {
        // if it's in the right subtree, remove it from that
        n.right.remove(n);
      } else {
        // if it's greater than this node's value but there's no right subtree, we don't have it
        console.log('value ' + n + ' was attempted to be removed from a tree that does not contain it');
        return false;
      }
    }

    return true;
  }

  // deletes the node it's called from
  // since we're not managing the memory ourselves,
  // this just means handling this node's children
  // and un-referencing it from the parent
  function deleteNode(node) {
    if(!node.left && !node.right) {
      // if this is a leaf we can just un-reference it

      // figure out which child of the parent this is
      // then remove it
      if(node.parent.value < node.value) {
        node.parent.right = null;
      } else {
        node.parent.left = null;
      }
    } else if((node.left || node.right) && !(node.left && node.right)) {
      // if this has only one child, take the value closest to the current and use that instead
      var descendant = node.left ? node.left.getLargestDescendant() : node.right.getSmallestDescendant();

      // move the descendant node here and
      // delete the old copy
      node.value = descendant.value;
      node.count = descendant.count;
      deleteNode(descendant);
    } else {
      // if we have two children, take the greatest valued node
      // from the left subtree and use that here
      var descendant = getLargestDescendant(node.left);

      // move the descendant node here and
      // delete the old copy
      node.value = descendant.value;
      node.count = descendant.count;
      deleteNode(descendant);
    }
  }

  function getSmallestDescendant(node) {
    return node.left ? node.left.getSmallestDescendant() : node;
  }

  function getLargestDescendant(node) {
    return node.right ? node.right.getLargestDescendant() : node;
  }

  function contains(i) {
    return findNode(this, i) == true;
  }

  function findNode(curr, i) {
    if(i === curr.value) {
      return curr;
    } else if(i < curr.value) {
      return curr.left ? curr.left.findNode(i) : null;
    } else {
      return curr.right ? curr.right.findNode(i) : null;
    }
  }

  // the height of a tree from node N is
  // the greatest of N's children's heights plus 1 for N itself
  function height() {
    var childHeight = 0;

    if(this.left) {
      childHeight = this.left.height();
    }

    if(this.right) {
      childHeight = Math.max(childHeight, this.right.height());
    }

    return childHeight + 1;
  }

  return {
    insert: insert,
    remove: remove,
    contains: contains,
    height: height
  }
}.call(this))
